ARCHITECTURE.md

```diff
@@ -0,0 +1,148 @@
+details. To that extent we're in the process of moving packages to `internal/`
+(and sometimes back to `pkg/`, if necessary).
+
+A note on semantic versioning: `gopass` is both an CLI and an API (Go module).
+The expectations around semantic versioning and Go modules make it hard express
+both concerns in the same versioning scheme, e.g. does a breaking change in the
+API require a major version bump even if nothing about the tool (CLI) has
+changed? What about the other way round? Thus we have decided to apply semantic
+versioning only to the CLI tool, not the Go module. This is not ideal and might
+change with sufficient active contributors.
+
+### `docs/backends`
+
+This folder contains documentation about each of our supported backends. See
+`internal/backend` below for more information about our backend design.
+
+### `docs/commands`
+
+This folder contains the specification of each subcommand the tool offers.
+We have many subcommands with sometimes dozens of flags each. In the past we
+did encounter some inconsistencies and decided to introduce specifications for
+each command. If the specification and the implementation disagree this should
+be reported as a bug a fixed or the specification needs to be changed (but the
+general assumptions should be that the specification is correct, not the code).
+
+### `docs/usecases`
+
+This directory contains an (incomplete) list of our core use cases, i.e. the
+critical user journeys we aim to support. `gopass` can be used in various ways
+and try to remain flexible and extensible, but if we encounter a conflict
+between a blessed use case and a corner case we're prefer the former.
+
+### `helpers/`
+
+This directory contains some release automation tooling that is supposed to be
+invoked with `go run`. The changelog generator in `helpers/changelog` is used
+by our GitHub Action based release automation and shouldn't be invoked manually.
+
+The tooling in `helpers/release` will prepare a new release and helps to file a
+release pull request will all the required updates in place.
+
+### `internal/` and `pkg/`
+
+`gopass` used to not have either of these and all our packages were rooted
+directly in the repository. However we began to notice that other projects
+were starting to depend directly on our internal packages and we sometimes
+broke them. This put both us and the other project into an unpleasant
+situation so we tried to clarify the expectations by using Gos `internal/`
+visbility rule to keep other projects from depending on our implementation
+details.
+
+Note: If we have a good reasons to use one of our `internal/` packages either
+copy it (our license should rarely be an issue) or nicely ask us and explain
+why something should move to `pkg/`.
+
+As we are in the process of formalizing a proper API surface we sometimes need
+to move packages from `internal/` to `pkg/`. The other direction might also
+occur, but much less often.
+
```

Comment:
Shouldn't we try at some point to split the CLI and the lib? And thus have all the pkg/ repo living in their lib repo with good semver versionning, maybe?

Reply:
Yes, but right now this would create too much work. I've hinted at this around line 35.
If we have enough active contributors we could consider that.